src/conn/conn_compact.c

```diff
@@ -21,21 +22,256 @@ __compact_server_run_chk(WT_SESSION_IMPL *session)
+    if (temp_compact_stat == NULL) {
+        WT_ERR(__wt_calloc_one(session, &temp_compact_stat));
+        WT_ERR(__wt_strdup(session, uri, &temp_compact_stat->uri));
+        __background_compact_list_insert(session, temp_compact_stat);
+    }
+
+    /* Fill starting information prior to running compaction. */
+    WT_ERR(bm->size(bm, session, &temp_compact_stat->start_size));
+    temp_compact_stat->prev_compact_time = __wt_clock(session);
+
+    *compact_stat = temp_compact_stat;
+
+    return (0);
+
+err:
+    __wt_free(session, temp_compact_stat);
+
+    return (ret);
+}
+
+/*
+ * __background_compact_end --
+ *     Fill resulting compact statistics in the background compact tracking list for a given file.
+ */
+static int
+__background_compact_end(WT_SESSION_IMPL *session, WT_BACKGROUND_COMPACT_STAT *compact_stat)
+{
+    WT_BM *bm;
+    WT_CONNECTION_IMPL *conn;
+    wt_off_t bytes_recovered;
+
+    bm = S2BT(session)->bm;
+    conn = S2C(session);
+
+    WT_RET(bm->size(bm, session, &compact_stat->end_size));
+    compact_stat->bytes_rewritten = bm->block->compact_bytes_rewritten;
+    bytes_recovered = compact_stat->start_size - compact_stat->end_size;
+
+    /*
+     * If the file failed to decrease in size, mark as an unsuccessful attempt. It's possible for
+     * compaction to do work (rewriting bytes) while other operations cause the file to increase in
+     * size.
+     */
+    if (bytes_recovered <= 0) {
+        compact_stat->consecutive_unsuccessful_attempts++;
+        compact_stat->prev_compact_success = false;
+    } else {
+        compact_stat->consecutive_unsuccessful_attempts = 0;
+        conn->background_compact.files_compacted++;
+        compact_stat->prev_compact_success = true;
+
+        /*
+         * Update the moving average of bytes rewritten across each file compact attempt. A
+         * weighting of 10% means that we are effectively considering the last 10 attempts in the
+         * average.
+         */
+        conn->background_compact.bytes_rewritten_ema =
+          (uint64_t)(0.1 * bm->block->compact_bytes_rewritten +
+            0.9 * conn->background_compact.bytes_rewritten_ema);
+    }
+
+    return (0);
+}
+
+/*
+ * __background_compact_list_cleanup --
+ *     Free all or any entry that has not been updated for more than a day in the background compact
+ *     tracking list.
+ */
+static void
+__background_compact_list_cleanup(
+  WT_SESSION_IMPL *session, WT_BACKGROUND_COMPACT_CLEANUP_TYPE cleanup_type)
+{
+    WT_BACKGROUND_COMPACT_STAT *compact_stat, *temp_compact_stat;
+    WT_CONNECTION_IMPL *conn;
+    uint64_t cur_time;
+
+    conn = S2C(session);
+    cur_time = __wt_clock(session);
+
+    TAILQ_FOREACH_SAFE(compact_stat, &conn->background_compact.compactqh, q, temp_compact_stat)
+    {
+        if (cleanup_type == BACKGROUND_CLEANUP_ALL_STAT ||
+          WT_CLOCKDIFF_SEC(cur_time, compact_stat->prev_compact_time) > WT_DAY)
```

Comment:
My intuition is that a day isn't a great granularity for this. The idea is that the background will be turned on during maintenance windows, which are likely to be once per day. I can see cases where this queue gets cleared out just before compact gets turned back on.

It is also going to be very difficult to test - since I suspect we don't want to run a test for a day to take this path.

Reply:
I agree, these conditions are hard to test in general since we're trying to handle long-running workloads. We hope to refine these as we do more testing. We could also do a debug_mode configurations here as you suggest.